Geometry/HGCalCommonData/src/HGCalWaferMask.cc

```diff
@@ -478,3 +478,239 @@ std::pair<int, int> HGCalWaferMask::getTypeMode(const double& xpos,
                                   << ":" << rotn;
   return ((mode == 0) ? std::make_pair(ncor, rotn) : std::make_pair(type, (rotn + HGCalWaferMask::k_OffsetRotation)));
 }
+
+bool HGCalWaferMask::goodTypeMode(
+    double xpos, double ypos, double delX, double delY, double rin, double rout, int part, int rotn, bool debug) {
+  if (part < 0 || part > HGCalTypes::WaferSizeMax)
+    return false;
+  if (rotn < 0 || rotn > HGCalTypes::WaferCornerMax)
+    return false;
+  static const int corners = HGCalTypes::WaferCornerMax;
+  static const int corner2 = 2 * HGCalTypes::WaferCornerMax;
+  static const int base = 10;
+  static const int base2 = 100;
+  double dx0[corners] = {0.0, delX, delX, 0.0, -delX, -delX};
+  double dy0[corners] = {-delY, -0.5 * delY, 0.5 * delY, delY, 0.5 * delY, -0.5 * delY};
+  double dx1[corners] = {0.5 * delX, delX, 0.5 * delX, -0.5 * delX, -delX, -0.5 * delX};
+  double dy1[corners] = {-0.75 * delY, 0.0, 0.75 * delY, 0.75 * delY, 0.0, -0.75 * delY};
+  double dx2[corner2] = {0.225 * delX,
+                         0.775 * delX,
+                         delX,
+                         delX,
+                         0.775 * delX,
+                         0.225 * delX,
+                         -0.225 * delX,
+                         -0.775 * delX,
+                         -delX,
+                         -delX,
+                         -0.775 * delX,
+                         -0.225 * delX};
+  double dy2[corner2] = {-0.8875 * delY,
+                         -0.6125 * delY,
+                         -0.275 * delY,
+                         0.275 * delY,
+                         0.6125 * delY,
+                         0.8875 * delY,
+                         0.8875 * delY,
+                         0.6125 * delY,
+                         0.275 * delY,
+                         -0.275 * delY,
+                         -0.6125 * delY,
+                         -0.8875 * delY};
+  bool ok(true);
+  int ncf(-1);
+  switch (part) {
+    case (HGCalTypes::WaferThree): {
+      static const int nc0[corners] = {450, 150, 201, 312, 423, 534};
+      int nc = nc0[rotn];
+      for (int k1 = 0; k1 < 3; ++k1) {
+        int k = nc % base;
+        double xc1 = xpos + dx0[k];
+        double yc1 = ypos + dy0[k];
+        double rpos = sqrt(xc1 * xc1 + yc1 * yc1);
```

Comment:
minor optimization: to avoid `sqrt` inside a loop, store `rout2 = rout*rout` and `rin2 = rin*rin` outside of the switch, then you can just compare to `rpos2 = xc1 * xc1 + yc1 * yc1` (for every case)